parquet-column/src/main/java/org/apache/parquet/schema/GroupType.java

```diff
@@ -349,6 +355,28 @@ void checkGroupContains(Type subType) {
     }
   }
 
+  /**
+   * Checks compatibility with given subType
+   * Allows subType to have extra optional fields
+   *
+   * @param subType
+   */
+  void checkGroupCompatibility(Type subType) {
+    if (subType.isPrimitive()) {
+      throw new InvalidRecordException(subType + " found: expected " + this);
+    }
+    List<Type> fields = subType.asGroupType().getFields();
+    for (Type otherType : fields) {
+      if(this.containsField(otherType.getName())) {
+        Type thisType = this.getType(otherType.getName());
+        thisType.checkCompatibility(otherType);
+      } else if (otherType.getRepetition() == Repetition.REQUIRED) {
```

Comment:
Unlike `checkContains` this throws an exception only when the extra field is `required`. Since, `optional` and `repeated` are 0 or 1 and 0 or more respectively, it's still compatible to have them as extra fields.